README.md
## An API that recovers data from the Wigor schedule with Playwright


Wigor schedule: a private schedule used in my school, written in .NET, and quite broken, I written a little comment about that, see in schedule.spec .  
(https://github.com/Theo-Dancoisne/theoslab.APIs.wigorSchedule/blob/fe36a860b69a66517607c9281fb7ef00ae7de638/e2e/schedule.spec.ts#L7-L18)  

As Wigor schedule uses personal login I invite you to read [./private/exemple.users.json](./private/example.users.json) .

#### To use this API you need either:  
- Your personal school login ;  
- Or use my private hosted version where possibly people from other promotions/specialties generously entrust me with their login that you can use through keys (ex: 3sys for 3rd year sysops option).  
  If you want more information about this contact me here 
  [theo.dancoisne@ecoles-epsi.net](mailto:theo.dancoisne@ecoles-epsi.net?subject=Je%20veux%20plus%20d'informations%20sur%20l'API%20Wigor%20schedule).  
  Or if you want to entrust me with your login, contact me here [theo.dancoisne@ecoles-epsi.net](mailto:theo.dancoisne@ecoles-epsi.net?subject=Je%20veux%20faire%20parti%20de%20l'API%20Wigor%20schedule&body=⚠Ne%20me%20donnez%20pas%20vos%20login%20maintenant⚠).  

##### Hosted version try out and more info on Postman
https://www.postman.com/theodancoisne/workspace/theo-s-lab-apis-public

### About Plywright

Playwright is used to do automation on browsers and apps  (headless or not) to check that everything works as expected.  
Here i use it to to grab the text in the html DOM.

[Playwright website](https://playwright.dev/)  
[Little Playwright doc by Microsoft](https://learn.microsoft.com/en-us/microsoft-edge/playwright/)

[Playwright VS Puppeteer](https://blog.logrocket.com/playwright-vs-puppeteer/)  

CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.2.1](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/compare/0.2.0...0.2.1)

- 📝 the changelog [`7842012`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/7842012a1b055b70ea18ac97d5b4ec95582b899f)

#### [0.2.0](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/compare/0.1.3...0.2.0)

> 5 November 2023

- updating packages [`5cf67ce`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/5cf67ce3ade03fc05ea8e18912652f1a85b86280)
- changing function call order and organization,  ➕ debugging tools [`715550a`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/715550aae558069b26efdc8bac7a0e7310aa08ff)
- replaced Selector_exists() by Selector_visible(), handles (hope) .NET error page [`b14326b`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/b14326b0005c9d6d30ff969ff741794baadaa2d1)

#### [0.1.3](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/compare/0.1.2...0.1.3)

> 22 October 2023

- exception handling: lack of Teams links [`f89ce1f`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/f89ce1f26306f974cb268ecb04977e6ce97f6049)
- Release 🏷0.1.3 [`7218648`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/7218648b7f9b946d0b2e5c2f37b5cf5cd18146a8)

#### [0.1.2](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/compare/0.1.1...0.1.2)

> 26 September 2023

- Release 🏷0.1.2 [`75a2939`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/75a293914771c15e4559c4e0db5661d7a2364bd4)
- push prod conf [`eedf3bd`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/eedf3bdf005ccf273fe9c44fc8f5f3da83056699)

#### [0.1.1](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/compare/0.1.0...0.1.1)

> 10 September 2023

- ➕ try/catch to print unhandled errors [`db36ca7`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/db36ca70f72cabeba8483e11fdbbf56a4777d58d)
- Release 🏷0.1.1 [`c7bd1a3`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/c7bd1a35d20128d8146366fe8e952a14cf2383f1)

#### 0.1.0

> 4 September 2023

- removing waitFor(...) timeout, ➕ release-it [`4265e60`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/4265e60b8c46c35574536162ebfb9925a41b2a7b)
- server test [`06438d6`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/06438d6892d79cbdd5f35329a697ff8dea9faef2)
- server test [`8a44302`](https://github.com-theoslab.apis.wigorschedule/Theo-Dancoisne/theoslab.APIs.wigorSchedule/commit/8a443025c24ff1c877fa99c30a117034e24c1681)


e2e/schedule.spec.ts
// I use this file to develop and test the API (index.js) quickly and gradually thanks to the VSCode extension 
// "Playwright Test for VSCode": https://marketplace.visualstudio.com/items?itemName=ms-playwright.playwright
// you can ignore it
import { test, expect, Page, Frame, Locator } from '@playwright/test';


/* why I (and people) hate the schedule of wigorservices: */
// - damn it's a long time to connect to `cas-p.wigorservices.net/cas/login` and then
//   to load the desired page on `ws-edt-cd.wigorservices.net/WebPsDyn.aspx`;
// - depending on the domain, you will comme across other versions of site which are still hosted and online ¯\_(°-°)_/¯ ;
// - the ?hash= url can sometimes lead you to the "Erreur de paramètres" (error of parameters) page ;
// - if not, you have still ~5% chances of this even if the url is correct, re-send the request without changing 
//   anything and you have still ~5% chances ;
// - events are not children of days, so I refer to the left position of the style
//   (which is the same to both types) with another matrix ;
// - it loads the header of the days of the week after and before the desired date, but not the events,
//   I don't understand the purpose of this but I have to uncount the 5 days before and after in my locator ;
/*  */

async function selector_exists(selector, timeout=5000) {
    try {
        return await selector.waitFor({ timeout: timeout, state: 'attached'});
    } catch (error) {
        return false;
    }
}

test("connect to wigor", async ({ page }) => {
    test.slow();
    const packageInfos = require("../package.json");
    const translateMonth = {
        janvier: 1,
        février: 2,
        mars: 3,
        avril: 4,
        mai: 5,
        juin: 6,
        juillet: 7,
        aout: 8,
        septembre: 9,
        octobre: 10,
        novembre: 11,
        décembre: 12,
    };
    var dateAtPos = {
        103: null,
        122.4: null,
        141.8: null,
        161.2: null,
        180.6: null,
    };
    var Events = {
        data: {},
        infos: {
            API_version: packageInfos.version,
            wigorSchedule_version: {},
        },
    };
    const users = require("../../private/wigor_schedule/.users.json");

    page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D10%252F23%252F2023%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    // page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D3%252F28%252F2023%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    // page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D12%252F30%252F2024%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    // page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D01%252F02%252F2025%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    // page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D01%252F7%252F2025%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    await page.getByLabel("username").fill(users["3sys"].unm);          // for some reasons does not work in playwright library
    await page.getByLabel("password").fill(users["3sys"].pswd);
    await page.getByRole("button", { name: "Login" }).click();          // "Login" is valid in playwright test but in playwright library the browser will use the language/location set by the system

    await page.waitForLoadState("networkidle");

    // ERROR: handle incorrect credentials
    if (!page.url().startsWith("https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx")) console.error("Incorrect credentials, the following code will fail.");
    // ERROR

    Events.infos.wigorSchedule_version = (await page.locator("#DivEntete_Version").textContent()).match(/EDT - (.*)/)[1];

    // date format in the url: m/d/y
    // month is 0 indexed, 0 = january  |  getDay() return number between 0 and 6, 0 = sunday (not monday!)
    const dateInUrl = page.url().match(/date=(.*)&/)[1].split("/").map((item) => parseInt(item));
    const dayInUrl = new Date(dateInUrl[2], dateInUrl[0] - 1, dateInUrl[1]);

    // EXCEPTION: the date in the url does not enable us to have the correct year between the 24 December and the 07 January
    const janDecException = (() => {
        const firstDayOfWeek = new Date(dayInUrl);
        firstDayOfWeek.setDate(firstDayOfWeek.getDate() - dayInUrl.getDay() + 1);   // +1 because otherwise it will give you the Sunday of the previous week at 00:00:00 and so the last day will be the Saturday at 00:00:00, no false but not what I want
        const lastDayOfWeek = new Date(firstDayOfWeek);
        lastDayOfWeek.setDate(firstDayOfWeek.getDate() + 6);

        return firstDayOfWeek.getFullYear() !== lastDayOfWeek.getFullYear();
    })();
    // EXCEPTION
    
    const days = page.locator(".Jour");
    let dateAtPosKeys = Object.keys(dateAtPos);
    // 5 because they load 5 day before and after the current week
    for (let i = 5; i < await days.count() - 5; i++) {
        const day = (await days.nth(i).locator(".TCJour").textContent()).toLowerCase().split(" ").slice(1);
        const TranslateMonth = translateMonth[day[1]];
        const currentMonth = dayInUrl.getMonth() + 1;
        dateAtPos[dateAtPosKeys[i - 5]] = {
            d: day[0].padStart(2, "0"),
            m: TranslateMonth.toString().padStart(2, "0"),
            y: (() => {
                if (janDecException) {
                    if (TranslateMonth == 1 && 1 == currentMonth || TranslateMonth == 12 && 12 == currentMonth) {
                        return dayInUrl.getFullYear();
                    } else return dayInUrl.getFullYear() + (TranslateMonth == 1 ? 1 : -1);
                } else return dayInUrl.getFullYear();
            })(),
        };
    }
    const events = page.locator(".Case");
    if (await events.first().textContent() == "Pas de cours cette semaine") {
        return {};
    } else {
        for (let i = 0; i < await events.count() - 1; i++) {
            let content = await events.nth(i).getByRole("table");
            
            let position = events.nth(i).getAttribute("style").then((style) => {
                // @ts-ignore
                style = style.replaceAll(" ", "").split(/[:;]/);
                // @ts-ignore
                return parseFloat(Math.fround(style[style.indexOf("left") + 1].slice(0, -1) - 0.12).toFixed(1));    // 0.12 is an offset
            });
            
            let Start = (await content.locator(".TChdeb").textContent()).split(" - ");
            let End = Start.pop();
            let DateStart = dateAtPos[await position].d + "/" + dateAtPos[await position].m + "/" + dateAtPos[await position].y + " " + Start;
            let DateEnd = dateAtPos[await position].d + "/" + dateAtPos[await position].m + "/" + dateAtPos[await position].y + " " + End;
            let Title = content.locator(".TCase").first().textContent();

            let Link = content.locator(".TCase").locator(".Teams").locator("a");
            if (await selector_exists(Link, 1000)) Link = Link.first().getAttribute("href");
            else Link = "";

            let Teacher = (await content.locator(".TCProf").innerText()).split("\n");
            let SchoolYear = Teacher.pop();

            // text content exemple: "Salle:<>(<>)"
            // my regex sucks because it leaves an empty string at both ends of the array so I added filter()
            let Room = (await content.locator(".TCSalle").textContent()).split(/Salle:|\(|\)/).filter(item => item);
            let Location = Room.pop();

            Events.data[DateStart] = {
                start: await DateStart,
                end: await DateEnd,
                title: await Title,
                link: await Link,
                teacher: await Teacher[0],
                schoolYear: await SchoolYear,
                room: await Room[0],
                location: await Location,
            }
        }
    }
    console.log(Events);
});


tests/schedule.spec.ts
import { test, expect } from '@playwright/test';

test("connect to wigor", async ({ page }) => {
    let url = "https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx?action=posEDTLMS&amp;serverID=C&amp;Tel=theo.dancoisne&amp;date=03/28/2023&amp;hashURL=6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2";
    await page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D03%252F28%252F2023%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    
    await page.getByLabel("username").fill("theo.dancoisne");
    await page.getByLabel("password").fill("DAN.the22.");
    await page.getByRole("button", { name: 'Login' }).click();

    await page.waitForURL(url);
    let a = await page.locator(".Case").all();
    await console.log(a);
});


tests/playwright.config.ts
// playwright.config.ts
import { PlaywrightTestConfig } from '@playwright/test';

const config: PlaywrightTestConfig = {
  projects: [
    {
      name: 'Microsoft Edge',
      use: {
        // Supported Microsoft Edge channels are: msedge, msedge-beta, msedge-dev, msedge-canary
        channel: 'msedge',
      },
    },
  ],
};

export default config

e2e/wigorSchedule.js
// const { test, expect } = require("@playwright/test");
const playwright = require("playwright");

(async () => {
    // const browser = await playwright.chromium.launchPersistentContext("C:/Users/theod/AppData/Local/Microsoft/Edge/User Data", {
    // const browser = await playwright.chromium.launchPersistentContext("C:/Users/theod/AppData/Local/Google/Chrome/User Data", {
    const browser = await playwright.chromium.launch({
        channel: 'msedge',
        // executablePath: "C:/Program Files (x86)/Microsoft/Edge/Application/msedge.exe",
        headless: false,
        // devtools: true,
        // handleSIGHUP: false,
        // handleSIGINT: false,
        // handleSIGTERM: false,
    });
    const context = await browser.newContext();
    const page = await context.newPage();

    let url = "https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx?action=posEDTLMS&amp;serverID=C&amp;Tel=theo.dancoisne&amp;date=03/28/2023&amp;hashURL=6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2";
    var scheduleObject = {};
    await page.goto("https://cas-p.wigorservices.net/cas/login?service=https%3A%2F%2Fws-edt-cd.wigorservices.net%2FWebPsDyn.aspx%3Faction%3DposEDTLMS%26serverID%3DC%26Tel%3Dtheo.dancoisne%26date%3D03%252F28%252F2023%26hashURL%3D6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2");
    await page.waitForLoadState("networkidle");
    console.log("ok");
    await page.getByLabel("username").fill("theo.dancoisne");
    // await page.getByLabel("password").fill("DAN.the22.");
    // await page.getByRole("button", { name: 'Login' }).click();

    // await page.waitForURL("https://ws-edt-cd.wigorservices.net/**");
    /*await page.waitForLoadState("networkidle");

    for (const element of await page.locator(".Case").all()) 
    {
        console.log(element.getByRole("table").locator(".TCase").textContent());
    }*/
    await browser.close();
})();

playwright.config.ts
import { defineConfig, devices } from '@playwright/test';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
export default defineConfig({
  testDir: './e2e',
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: 'html',
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://127.0.0.1:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
  },

  /* Configure projects for major browsers */
  projects: [
    // {
    //   name: 'chromium',
    //   use: { ...devices['Desktop Chrome'] },
    // },

    // {
    //   name: 'firefox',
    //   use: { ...devices['Desktop Firefox'] },
    // },

    // {
    //   name: 'webkit',
    //   use: { ...devices['Desktop Safari'] },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: { ...devices['Pixel 5'] },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: { ...devices['iPhone 12'] },
    // },

    /* Test against branded browsers. */
    {
      name: 'Microsoft Edge',
      use: { ...devices['Desktop Edge'], channel: 'msedge' },
    },
    // {
    //   name: 'Google Chrome',
    //   use: { ...devices['Desktop Chrome'], channel: 'chrome' },
    // },
  ],

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   url: 'http://127.0.0.1:3000',
  //   reuseExistingServer: !process.env.CI,
  // },
});


index.js
const playwright = require("playwright");
const packageInfos = require("./package.json");

class WigorSchedule {
    #users;
    #browser;
    #context;
    #page;
    headless = true;
    translateMonth = {
        janvier: 1,
        février: 2,
        mars: 3,
        avril: 4,
        mai: 5,
        juin: 6,
        juillet: 7,
        aout: 8,
        septembre: 9,
        octobre: 10,
        novembre: 11,
        décembre: 12,
    };
    dateAtPos = {
        103: null,
        122.4: null,
        141.8: null,
        161.2: null,
        180.6: null,
    };
    Events = {
        data: {},
        infos: {
            API_version: packageInfos.version,
            wigorSchedule_version: null,
            request_parameters: {
                method: null,
                timestamp: null,
                profile_name: null,
                user_name: null,
            },
        },
        error: {},
    };
    EventsCopy = JSON.parse(JSON.stringify(this.Events));

    constructor(path = "../private/.users.json", headless = true) {
        try {
            this.#users = require(path);
            this.headless = headless;
        } catch (error) {
            throw new Error(`${error} \nYou may need to check out at ./private/example.users.json .\n`);
        }
    }

    /* browser context setup */
    async SetBrower() {
        try {
            if (this.headless) {
                this.#browser = await playwright.chromium.launch();      // prod, headless
            }
            else {
                this.#browser = await playwright.chromium.launch({       // dev, headed
                    channel: 'msedge',
                    headless: false,
                });
            }
            this.#context = await this.#browser.newContext();
            this.#page = await this.#context.newPage();
        }
        catch(error) {
            throw new Error(`${error} \nWhile setting up the Playwright browser, did you forget to install the correct one ? Headless was set on ${headless}, is that what you intended ?`);
        }
    }
    /*  */

    GetEventsByProfile(userProfile, date = new Date()) {
        this.Events.infos.request_parameters.method = "GET";
        this.Events.infos.request_parameters.profile_name = userProfile;
        return new Promise((resolve, reject) => {
            if (!this.#users[userProfile]) reject(`The profile '${userProfile}' does not exist.`);
            this.Start(this.#users[userProfile].unm, this.#users[userProfile].pswd, date).then(() => {
                this.#browser.close();
                resolve(this.Events);
            },
            (err) => {
                let datetime = new Date();
                this.CaptureError(datetime)
                .catch((error) => console.log(error))
                .then(() => this.#browser.close());
                this.Events.error = `[${datetime.toUTCString()}] ${err}`;
                reject(this.Events);
            });
        });
    }
    GetEvents(username, password, date = new Date()) {
        this.Events.infos.request_parameters.method = "POST";
        this.Events.infos.request_parameters.user_name = username;
        return new Promise((resolve, reject) => {
            this.Start(username, password, date).then(() => {
                this.#browser.close();
                resolve(this.Events);
            },
            (err) => {
                let datetime = new Date();
                this.CaptureError(datetime)
                .catch((error) => console.log(error))
                .then(() => this.#browser.close());
                this.Events.error = `[${datetime.toUTCString()}] ${err}`;
                reject(this.Events);
            });
        });
    }

    ClearEventsproperty() {
        this.Events = JSON.parse(JSON.stringify(this.EventsCopy));
    }

    async Selector_visible(selector, timeout=5000) {
        try {
            return await selector.isVisible({ timeout: timeout});
        } catch (error) {
            return false;
        }
    }

    CaptureError(datetime) {
        return new Promise((resolve, reject) => {
            this.#page.screenshot({
                path: `theoslab.APIs.wigorSchedule/logs/screenshots/error_${datetime.getDate()}-${datetime.getMonth() + 1}-${datetime.getFullYear()}_${datetime.getHours()}-${datetime.getMinutes()}-${datetime.getMilliseconds()}.png`,
                timeout: 5000
            }).then(() => {
                return resolve();
            }
            ,(error) => {
                return reject(`Error while taking screenshot:\n${error}`);
            });
        });
    }

    async Start(username, password, date) {
        await this.SetBrower().then(() => {},
        (err) => {
            return reject(err);
        });
        
        this.Events.infos.request_parameters.timestamp = date.getTime();
        // https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx?action=posEDTLMS&serverID=C&Tel=theo.dancoisne&date=02/28/2023&hashURL=6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2
        let url = [
            "https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx?action=posEDTLMS&serverID=C&Tel=",
            username,
            "&date=",
            (date.getMonth() + 1) + "/" + date.getDate() + "/" + date.getFullYear(),
            "&hashURL=6A322522A712EBD110260D1D505E28F595156D9701C0D240D268F8F329899514AFCCC45DAE4C54C0329C0765F10306871431A8FDA76A5C561114CD87028866D2"
        ]
        
        this.#page.goto(url.join(""));

        return new Promise(async (resolve, reject) => {
            // await this.#page.getByLabel("username").fill(username);       // for some reasons does not work outside of playwright test
            await this.#page.locator("#username").fill(username);
            await this.#page.getByLabel("password").fill(password);
            // await this.#page.getByRole("button", { name: "Login" }).click();          // "Login" is valid in playwright test but in playwright library the browser will use the language/location set by the system
            await this.#page.locator("[name=submitBtn]").click();
    
            await this.#page.waitForLoadState("networkidle", {timeout: 0});

            // ERROR: handle incorrect credentials
            if (!this.#page.url().startsWith("https://ws-edt-cd.wigorservices.net/WebPsDyn.aspx")) {
                return reject("Incorrect username or password for Wigor Services.\nIf you see this error while using a profile on my web hosted version, please report this issue on https://github.com/Theo-Dancoisne/theoslab.APIs.wigorSchedule/issues/new?title=Nom%20d%27utilisateur%20ou%20mot%20de%20passe%20incorrecte%20dans%20le%20profile&body=Indiquez%20le%20nom%20du%20profile%20que%20vous%20avez%20utilisé..");
            }
            // ERROR

            // ERROR: wrong page after redirection by login(success) page (e.g: .NET error page) OR Unhandled Exception!
            let version = this.#page.locator("#DivEntete_Version");
            if (await this.Selector_visible(version)) this.Events.infos.wigorSchedule_version = (await version.textContent()).match(/EDT - (.*)/)[1];
            else {
                return reject("Wigor schedule version not found, this means that the login page redirected to the wrong page (e.g: a .NET error page of the app) or that the schedule page has been updated.\nTry once or twice before reporting the issue.");
            }
            // ERROR

            // return reject("this a test");
            // date format in the url: m/d/y
            // month is 0 indexed, 0 = january  |  getDay() return number between 0 and 6, 0 = sunday (not monday!)
            const dateInUrl = this.#page.url().match(/date=(.*)&/)[1].split("/").map((item) => parseInt(item));
            const dayInUrl = new Date(dateInUrl[2], dateInUrl[0] - 1, dateInUrl[1]);
    
            // EXCEPTION: the date in the url does not enable us to have the correct year between the 24 December and the 07 January
            const janDecException = (() => {
                const firstDayOfWeek = new Date(dayInUrl);
                firstDayOfWeek.setDate(firstDayOfWeek.getDate() - dayInUrl.getDay() + 1);   // +1 because otherwise it will give you the Sunday of the previous week at 00:00:00 and so the last day will be the Saturday at 00:00:00, no false but not what I want
                const lastDayOfWeek = new Date(firstDayOfWeek);
                lastDayOfWeek.setDate(firstDayOfWeek.getDate() + 6);
    
                return firstDayOfWeek.getFullYear() !== lastDayOfWeek.getFullYear();
            })();
            // EXCEPTION
            
            try {
                const days = this.#page.locator(".Jour");
                let dateAtPosKeys = Object.keys(this.dateAtPos);
                // 5 because they load 5 day before and after the current week
                for (let i = 5; i < await days.count() - 5; i++) {
                    const day = (await days.nth(i).locator(".TCJour").textContent()).toLowerCase().split(" ").slice(1);
                    const TranslateMonth = this.translateMonth[day[1]];
                    const currentMonth = dayInUrl.getMonth() + 1;
                    this.dateAtPos[dateAtPosKeys[i - 5]] = {
                        d: day[0].padStart(2, "0"),
                        m: TranslateMonth.toString().padStart(2, "0"),
                        y: (() => {
                            if (janDecException) {
                                if (TranslateMonth == 1 && 1 == currentMonth || TranslateMonth == 12 && 12 == currentMonth) {
                                    return dayInUrl.getFullYear();
                                } else return dayInUrl.getFullYear() + (TranslateMonth == 1 ? 1 : -1);
                            } else return dayInUrl.getFullYear();
                        })(),
                    };
                }
            } catch (error) {
                return reject(`Unhandled exception when retrieving days of the week, if you think it should be reported, do so on https://github.com/Theo-Dancoisne/theoslab.APIs.wigorSchedule/issues/new/choose.\nerror: \n ${error}`);
            }

            try {
                const events = this.#page.locator(".Case");
                if (await events.first().textContent() !== "Pas de cours cette semaine") {
                    for (let i = 0; i < await events.count() - 1; i++) {
                        let content = await events.nth(i).getByRole("table");
                        
                        let position = events.nth(i).getAttribute("style").then((style) => {
                            style = style.replaceAll(" ", "").split(/[:;]/);
                            return parseFloat(Math.fround(style[style.indexOf("left") + 1].slice(0, -1) - 0.12).toFixed(1));    // 0.12 is an offset
                        });
                        
                        let Start = (await content.locator(".TChdeb").textContent()).split(" - ");
                        let End = Start.pop();
                        let DateStart = this.dateAtPos[await position].d + "/" + this.dateAtPos[await position].m + "/" + this.dateAtPos[await position].y + " " + Start;
                        let DateEnd = this.dateAtPos[await position].d + "/" + this.dateAtPos[await position].m + "/" + this.dateAtPos[await position].y + " " + End;
                        let Title = content.locator(".TCase").first().textContent();

                        let Link = content.locator(".TCase").locator(".Teams").locator("a").first();
                        if (await this.Selector_visible(Link, 1000)) Link = Link.getAttribute("href");
                        else Link = "";

                        let Teacher = (await content.locator(".TCProf").innerText()).split("\n");
                        let SchoolYear = Teacher.pop();
            
                        // text content exemple: "Salle:<>(<>)"
                        // my regex sucks because it leaves an empty string at both ends of the array so I added filter()
                        let Room = (await content.locator(".TCSalle").textContent()).split(/Salle:|\(|\)/).filter(item => item);
                        let Location = Room.pop();
            
                        this.Events.data[DateStart] = {
                            start: await DateStart,
                            end: await DateEnd,
                            title: await Title,
                            link: await Link,
                            teacher: await Teacher[0],
                            schoolYear: await SchoolYear,
                            room: await Room[0],
                            location: await Location,
                        }
                    }
                }
                return resolve();
            } catch (error) {
                return reject(`Unhandled exception when retrieving the events, if you think it should be reported, do so on https://github.com/Theo-Dancoisne/theoslab.APIs.wigorSchedule/issues/new/choose.\nerror:\n${error}`);
            }
        });
    }
}

exports.WigorSchedule = WigorSchedule;
